contracts/TokenClaim.sol

```diff
@@ -0,0 +1,65 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+import "./interfaces/ITokenClaim.sol";
+import "./VotesToken.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract TokenClaim is FactoryFriendly, ITokenClaim {
+    using SafeERC20 for IERC20;
+
+    address public childToken;
+    address public parentToken;
+    uint256 public snapShotId;
+    uint256 public parentAllocation;
+    mapping(address => bool) claimed;
+
+    /// @notice Initialize function, will be triggered when a new proxy is deployed
+    /// @param initializeParams Parameters of initialization encoded
+    function setUp(bytes memory initializeParams) public override initializer {
+        __Ownable_init();
+        (
+            address _metaFactory,
+            address _parentToken,
+            address _childToken,
+            uint256 _parentAllocation
+        ) = abi.decode(initializeParams, (address, address, address, uint256));
+
+        childToken = _childToken;
+        parentToken = _parentToken;
+        parentAllocation = _parentAllocation;
+
+        snapShotId = VotesToken(_parentToken).captureSnapShot();
+
+        IERC20(_childToken).transferFrom(_metaFactory, address(this), _parentAllocation);
+
+        emit SnapShotCaptured(_parentToken, _childToken, _parentAllocation);
+    }
+
+    /// @notice This function allows pToken holders to claim cTokens
+    /// @param claimer Address which is being claimed for
+    function claimToken(address claimer) external {
+        uint256 amount = getClaimAmount(claimer); // Get user balance
+
+        if (amount == 0) revert NoAllocation();
+
+        claimed[claimer] = true;
+
+        IERC20(childToken).safeTransfer(claimer, amount); // transfer user balance
+
+        emit TokenClaimed(parentToken, childToken, claimer, amount);
+    }
+
+    /// @notice Gets a users child token claimable amount
+    /// @param claimer Address which is being claimed for
+    /// @return cTokenAllocation Users cToken allocation
+    function getClaimAmount(address claimer)
+        public
+        view
+        returns (uint256)
+    {
+        return claimed[claimer] ? 0 :
+            (VotesToken(parentToken).balanceOfAt(claimer, snapShotId) * parentAllocation) /
+            VotesToken(parentToken).totalSupplyAt(snapShotId);
```

Comment:
If the total supply of the parent token is zero, yes.
We could add in a require statement in the setUp function that checks that the parent token total supply is greater than zero, but this feels a bit overkill to me.

Reply:
Why does that feel overkill to you?

Reply:
It just seems to me that if the parent token total supply is zero, then they've got bigger issues than this function reverting. There's lots of ways a DAO could be setup that don't make sense / brick the DAO, and I don't think we can protect against all of them in the contracts.

Reply:
Can they change their token supply after creation with a proposal? In that case then maybe it makes sense to check, but otherwise seems odd to account for the case where it could be zero?

Reply:
yeah, or burning. Which to your point though if they vote to brick their DAO I'm not sure we should or can account for all the ways they can do it

Reply:
I can't think of any ways to abuse setting your total supply to zero here either

Reply:
Yeah agreed, I definitely don't think there's any way this could be used for abuse / manipulation